src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -6565,8 +6565,13 @@ and TcRecordConstruction (cenv: cenv) (overallTy: TType) env tpenv withExprInfoO
     let ns1 = NameSet.ofList (List.map fst fldsList)
     let ns2 = NameSet.ofList (List.map (fun x -> x.rfield_id.idText) fspecs)
 
-    if withExprInfoOpt.IsNone && not (Zset.subset ns2 ns1) then
-        error (MissingFields(Zset.elements (Zset.diff ns2 ns1), m))
+    match withExprInfoOpt with
+    | None ->
+        if not (Zset.subset ns2 ns1) then
+            error(MissingFields(Zset.elements (Zset.diff ns2 ns1), m))
+    | _ ->
+        if oldFldsList.IsEmpty && g.langVersion.SupportsFeature LanguageFeature.WarningWhenCopyAndUpdateRecordChangesAllFields then
+            warning(Error(FSComp.SR.tcCopyAndUpdateRecordChangesAllFields(fullDisplayTextOfTyconRef tcref), m))
```

Comment:
No. Is there an example where it works like that?

Reply:
Good question, I will check

Reply:
Guess we'd have to start conditionally disabling warnings based on language features in CompilerDiagnostics.fs?

Reply:
Yeah, that's what I had in mind, I wonder if we can do the follwing in CompilerDiagnostics:
```fsharp
    member x.IsEnabled(severity, options) =
//...
        | 3560 -> g.langVersion.SupportsFeature LanguageFeature.WarningWhenCopyAndUpdateRecordChangesAllFields // tcCopyAndUpdateRecordChangesAllFields - off by default
```
Probably will need some additional wiring to get the tcglobals there?

Wdyt @kerams?

Reply:
This way we will be able to promote certain warnings to "on by default" in new versions, but left intact for previous.

Reply:
Ugh, so TcGlobals is inaccessible there, and LanguageFeatures is internal, so it can't be stored in CompilerDiagnostics. I added it to `IsEnabled`, but now it needs to be passed in through about a billion calls, so I'm not sure it's the greatest idea.

Reply:
Maybe if `FSharpDiagnosticOptions` was transformed into a class, features could be exposed via a method.

Reply:
You could do a new type for the warning message, pass e.g. the current langVersion and needed feature next to it.
The warn() function expects an exception, so that should work.

PhasedDiagnostic.IsEnabled has then access to the exception, so you could unwrap it and react on it.


So instead of doing an "if" around the language feature, it will be passed together with the exception.

Reply:
Added `exception DiagnosticEnabledWithLanguageFeature of number: int * message: string * range: range * enabledByLangFeature: bool`. I could not think of a reason to add language version and required features separately instead of that bool, but I can change it if you want.

Reply:
No, it's equivalent - it would just a matter of where the logic of comparing versions/features should be. My original thinking was to pass the decision to phaseddiagnostics, but what you have is basically the same.